Mathlib/Algebra/Module/FinitePresentation.lean

```diff
@@ -0,0 +1,285 @@
+Authors: Andrew Yang
+-/
+import Mathlib.RingTheory.Noetherian
+import Mathlib.Algebra.Module.LocalizedModule
+import Mathlib.LinearAlgebra.Isomorphisms
+import Mathlib.LinearAlgebra.FreeModule.Finite.Basic
+/-!
+
+# Finitely Presented Modules
+
+## Main definition
+
+- `Module.FinitePresentation`: A module is finitely presented if it is finitely generated by some
+  set `s` and the kernel of the presentation `Rˢ → M` is also finitely generated.
+
+## Main results
+
+- `Module.finitePresentation_iff_finite`: If `R` is noetherian, then f.p. iff f.g. on `R`-modules.
+
+Suppose `0 → K → M → N → 0` is an exact sequence of `R`-modules.
+
+- `Module.finitePresentation_of_surjective`: If `M` is f.p., `K` is f.g., then `N` is f.p.
+
+- `Module.FinitePresentation.fg_ker`: If `M` is f.g., `N` is f.p., then `K` is f.g.
+
+- `Module.finitePresentation_of_ker`: If `N` and `K` is f.p., then `M` is also f.p.
+
+- `Module.FinitePresentation.isLocalizedModule_map`: If `M` and `N` are `R`-modules and `M` is f.p.,
+  and `S` is a submonoid of `R`, then `Hom(Mₛ, Nₛ)` is the localization of `Hom(M, N)`.
+
+
+Also the instances finite + free => f.p. => finite is also provided
+
+## TODO
+Suppose `S` is an `R`-algebra, `M` is an `S`-module. Then
+1. If `S` is f.p., then `M` is `R`-f.p. implies `M` is `S`-f.p.
+2. If `S` is both f.p. (as an algebra) and finite (as a module),
+  then `M` is `S`-fp implies that `M` is `R`-f.p.
+3. If `S` is f.p. as a module, then `S` is f.p. as an algebra.
+In particular,
+4. `S` is f.p. as an `R`-module iff it is f.p. as an algebra and is finite as a module.
+
+
+For finitely presented algeras, see `RingTheory/FinitePresentation`.
+-/
+
+variable (R M N) [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+/--
+A module is finitely presented if it is finitely generated by some set `s`
+and the kernel of the presentation `Rˢ → M` is also finitely generated.
+-/
+class Module.FinitePresentation : Prop where
+  out : ∃ (s : Finset M), Submodule.span R (s : Set M) = ⊤ ∧
+    (LinearMap.ker (Finsupp.total s M R Subtype.val)).FG
+
+instance (priority := 100) [h : Module.FinitePresentation R M] : Module.Finite R M := by
+  obtain ⟨s, hs₁, _⟩ := h
+  exact ⟨s, hs₁⟩
+
+-- Ideally this should be an instance but it makes mathlib much slower.
+lemma Module.finitePresentation_of_finite [IsNoetherianRing R] [h : Module.Finite R M] :
+    Module.FinitePresentation R M := by
+  obtain ⟨s, hs⟩ := h
+  exact ⟨s, hs, IsNoetherian.noetherian _⟩
+
+lemma Module.finitePresentation_iff_finite [IsNoetherianRing R] :
+    Module.FinitePresentation R M ↔ Module.Finite R M :=
+  ⟨fun _ ↦ inferInstance, fun _ ↦ finitePresentation_of_finite R M⟩
+
+variable {R M N}
+
+lemma Module.finitePresentation_of_free_of_surjective [Module.Free R M] [Module.Finite R M]
+    (l : M →ₗ[R] N)
+    (hl : Function.Surjective l) (hl' : Submodule.FG (LinearMap.ker l)) :
+    Module.FinitePresentation R N := by
+  classical
```

Comment:
Shouldn't code be indented two further spaces after `classical`? Looking through mathlib it seems that the vast majority of `classical`s have this property. Also on line 122.